docs/adr/002.md

```diff
@@ -0,0 +1,118 @@
+  relativistic Breit-Wigner that is applied to the resonance in some 1-to-3
+  isobar decay (described by an `STG` with final state edges 2, 3, 4 and
+  intermediate edge 1) would work on the invariant mass of edge 3 and 4
+  (`mSq_3+4`).
+- Just like variables, parameters need to be identifiable from their position
+  within the `STG` (take a relativistic Breit-Wigner _with form factors_, which
+  would require break-up momentum as a parameter), but also require some
+  suggested starting value (e.g. expected pole position). These starting values
+  are usually taken from the edge and node properties within the `STG`.
+
+(requirements-adr-002)=
+
+### Solution requirements
+
+1. It is easy to apply dynamics to specific components of the `STG`s.
+2. The parameters that a dynamics functions requires are registered
+   automatically and linked together.
+3. Where possible, suggested (initial) parameter values are provided as well.
+4. Kinematic variables used in dynamics functions are also linked
+   appropriately.
+5. It is easy to define custom dynamics and insert those into the amplitude
+   model.
+
+## Decision Drivers
+
+- [#124](https://github.com/ComPWA/expertsystem/issues/124): see
+  {ref}`adr/002:Issues with existing set-up`.
+- [#440](https://github.com/ComPWA/expertsystem/issues/440): no way to supply
+  custom dynamics. Or at least, {mod}`tensorwaves` does not understand those
+  custom dynamics.
+- [ADR-001](./001.md): parameters _and_ variables are to be expressed as
+  `sympy.Symbol`s.
+- [#454](https://github.com/ComPWA/expertsystem/pull/454): dynamics are
+  specified as a mapping of `sympy.Function` to a `sympy.Expr`, but now there
+  is no way to supply those `sympy.Expr`s with expected `sympy.Symbol`s
+  (parameters and variables).
+
+## Considered Options
+
+To satisfy [**requirement 1**](requirements-adr-002), we propose the following
+syntax:
+
+```python
+# model: ModelInfo
+# graph: StateTransitionGraph
+model.set_dynamics(graph, edge_id=1, expression_builder)
+```
+
+:::{toggle}
+
+Another style would be to have `ModelInfo` contain a reference to the list of
+`StateTransitionGraph`s. The user then needs some other way to express which
+edges to apply the dynamics function to:
+
+```python
+model.set_dynamics(
+    filter=lambda p: p.name.startswith("f(0)"),
+    edge_id=1,
+    expression_builder,
+)
+```
+
+:::
+
+Here, `expression_builder` is some function or method that can create a
+dynamics expression. It can also be a class that contains both the
+implementation of the expression and a static method to build itself from a
+`StateTransitionGraph`.
+
+The dynamics expression needs to be formulated in such a way that it satisfies
+[**requirements 2-5**](requirements-adr-002). The following options illustrate
+three different ways of formulating a dynamics expression, taking a
+relativistic Breit-Wigner and a relativistic Breit-Wigner _with form factor_ as
+example.
+
+```{toctree}
+002/composition
```

Comment:
Im adding my comment here since its a bit hard to find the right places in that notebook. So a couple of comments regarding the cons of the composition solutions:

* Regarding con 1. Why is the expression signature of the parameter dataset form? The expression is just that of a relativistic breit wigner or whatever function is implemented.
* Regarding con 2. I don't understand which argument number checking means here. The function defines what it needs and this has to be mapped with what is available. And at that point the checking can be done. But this as to be done no matter which solution is chosen. I dont really see how this check is automatically guaranteed.
* Regarding con 3. The function code can be included in the builder, just like the other solutions do that as well.
* Regarding con 4. Apart from a `Protocol`, abc / an interface can be defined and enforced.

Reply:
> its a bit hard to find the right places in that notebook

Ok, 9d7adf0 should make this easier. The idea is that the notebooks are only to illustrate implementation and usage, while the main Markdown ADR file is where discussion takes place.

> Regarding con 1. Why is the expression signature of the parameter dataset form? The expression is just that of a relativistic breit wigner or whatever function is implemented.

Could you clarify?

> Regarding con 2. (...) But this as to be done no matter which solution is chosen.

That's true, 3126617 rewrites the Pros and Cons a bit.

> Regarding con 3. The function code can be included in the builder, just like the other solutions do that as well.

I would have to see. I assume that you mean a function is returned from the builder. But that means there is only a builder function that takes an `STG`, so there's no way to use this function independently of the rest of the `expertsystem`.

> Regarding con 4. Apart from a Protocol, abc / an interface can be defined and enforced.

For functions, I think `Protocol` is the only way to enforce the function signature (or `Callable`, but that's just a slimmer version). But I'm not sure if that's a bad thing, see:
https://www.python.org/dev/peps/pep-0544/#rationale-and-goals

Reply:
Was discussed in a meeting. The modified adr will address these points